### PR TITLE
Directive instance refactor

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
> core/todo/bundle.br decreased by 4 bytes.

Yay :-)

File: packages/core/src/render3/interfaces/node.ts [hunk 1]
```diff
@@ -602,6 +602,11 @@ export interface TProjectionNode extends TNode {
   projection: number;
 }
 
+/**
+ * An union type representing all TNode types that can host a directive.
```
Comment (kara) [inline packages/core/src/render3/interfaces/node.ts]: :)

```suggestion
 * A union type representing all TNode types that can host a directive.
```

File: packages/core/src/render3/instructions/shared.ts [hunk 2]
```diff
@@ -1096,14 +1096,27 @@ function instantiateAllDirectives(tView: TView, lView: LView, tNode: TDirectiveH
...
+
   for (let i = start; i < end; i++) {
     const def = tView.data[i] as DirectiveDef<any>;
-    if (isComponentDef(def)) {
-      assertNodeOfPossibleTypes(tNode, TNodeType.Element);
-      addComponentLogic(lView, tNode as TElementNode, def);
+    const isComponent = isComponentDef(def);
+
+    if (isComponent) {
+      ngDevMode && assertNodeOfPossibleTypes(tNode, TNodeType.Element);
```
Comment (kara) [inline packages/core/src/render3/instructions/shared.ts]: Good catch!

**Review by kara — APPROVED** [review]
LGTM

**Comment by kara** [comment]
[VE presubmit](https://test.corp.google.com/ui#id=OCL:277752977:BASE:277753033:1572543532404:335b2e70)

[Ivy presubmit](http://test/OCL:277752977:BASE:277753512:1572543634097:cd3b95e9)

**Pull request closed without merging**